consensus/src/payload_manager.rs

```diff
@@ -1,119 +1,192 @@
-impl PayloadManager for QuorumStoreClient {
-    async fn pull_payload(
-        &self,
-        max_items: u64,
-        max_bytes: u64,
-        exclude_payloads: PayloadFilter,
-        wait_callback: BoxFuture<'static, ()>,
-        pending_ordering: bool,
-    ) -> Result<Payload, QuorumStoreError> {
-        fail_point!("consensus::pull_payload", |_| {
-            Err(anyhow::anyhow!("Injected error in pull_payload").into())
-        });
-        let mut callback_wrapper = Some(wait_callback);
-        // keep polling QuorumStore until there's payloads available or there's still pending payloads
-        let mut count = self.poll_count;
-        let payload = loop {
-            count -= 1;
-            let payload = self
-                .pull_internal(max_items, max_bytes, exclude_payloads.clone())
-                .await?;
-            if payload.is_empty() && !pending_ordering && count > 0 {
-                if let Some(callback) = callback_wrapper.take() {
-                    callback.await;
+    /// Extract transaction from a given block
+    /// Assumes it is never called for the same block concurrently. Otherwise status can be None.
+    pub async fn get_transactions(&self, block: &Block) -> Result<Vec<SignedTransaction>, Error> {
+        if block.payload().is_none() {
+            return Ok(Vec::new());
+        }
+
+        match self {
+            PayloadManager::DirectMempool => match block.payload().unwrap() {
+                Payload::InQuorumStore(_) => unreachable!("should not use QuorumStore"),
+                Payload::DirectMempool(txns) => Ok(txns.clone()),
+            },
+            PayloadManager::InQuorumStore(batch_reader, _) => {
+                match block.payload().unwrap() {
+                    Payload::DirectMempool(_) => unreachable!("should use QuorumStore"),
+                    Payload::InQuorumStore(proof_with_data) => {
+                        let status = proof_with_data.status.lock().take();
+                        match status.expect("Should have been updated before") {
+                            DataStatus::Cached(data) => {
+                                proof_with_data
+                                    .status
+                                    .lock()
+                                    .replace(DataStatus::Cached(data.clone()));
+                                Ok(data)
+                            }
+                            DataStatus::Requested(receivers) => {
+                                let mut vec_ret = Vec::new();
+                                debug!("QSE: waiting for data on {} receivers", receivers.len());
+                                for (digest, rx) in receivers {
+                                    match rx.await {
+                                        Err(_) => {
+                                            // We probably advanced epoch already.
+                                            warn!("Oneshot channel to get a batch was dropped");
```

Comment:
For `warn` messages, please add any relevant data that could help understand what has happened. E.g., at least the error message.